manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,275 @@
+package deallocator
+
+import (
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
```

Comment:
can we do map[interface{}]resourceDeallocator ? @wk8

Reply:
I'm not sure I understand the idea of not making it a bit more specific (reflect.Type than just an empty interface) if we can?

Reply:
Is reflect.Type more specific than interface{} ?

Reply:
Well yes, it's a well-defined type, as opposed to `interface{}` which matches anything, akin to C/C++'s `void*`?

Reply:
Thanks, interesting read.

The main argument against reflection here is performance (type safety is not a concern here, given the very controlled fashion in which it's used). And sure enough, it's slower than a case switch (see below). But at least that allows us to not hardcode in the package's main file the different resource types, and keep each resource cleanly separated from the deallocator's code.

If we wanted to keep that separation of concerns in the code, we could indeed use type assertion in a linked list of resource-specific implementations, but that might grow slow if we add a lot of resources (10+, according to the benchmark below).

A rough benchmark:
```
package foo

import (
	"reflect"
	"testing"
)

type Foo struct{}
type Bar struct{}

var a = []interface{}{Foo{}, Bar{}}

func BenchmarkTypeOf(b *testing.B) {
	foo := reflect.TypeOf(Foo{})
	bar := reflect.TypeOf(Bar{})

	x, y := 0, 0

	for n := 0; n < b.N; n++ {
		for _, i := range a {
			switch reflect.TypeOf(i) {
			case foo:
				x++
			case bar:
				y++
			}
		}
	}
}

func BenchmarkSwitch(b *testing.B) {
	x, y := 0, 0

	for n := 0; n < b.N; n++ {
		for _, i := range a {
			switch i.(type) {
			case Foo:
				x++
			case Bar:
				y++
			}
		}
	}
}

func BenchmarkTypeAssertion(b *testing.B) {
	x, y := 0, 0

	for n := 0; n < b.N; n++ {
		for _, i := range a {
			if _, isFoo := i.(Foo); isFoo {
				x++
			} else if _, isBar := i.(Bar); isBar {
				y++
			}
		}
	}
}
```
yields
```
BenchmarkTypeOf-8          	100000000	        20.9 ns/op
BenchmarkSwitch-8          	1000000000	         2.89 ns/op
BenchmarkTypeAssertion-8   	2000000000	         1.89 ns/op
```

so, happy to re-write as a list of type assertions if you want?

Reply:
(also, we only do a call to `reflect.Type` when there's an update to a network (or another service-level resource down the line), so the 10 ns price might be okay?)

Reply:
oh boy

Reply:
oh lord this is over-engineered.